Only resize images (cache) for themes which are being used instead of all installed themes

denisristic opened the pull request:
Only resize images (cache) for themes which are being used instead of all installed themes

### Description
Pull request is using [@vrann fix](https://github.com/magento/magento2/pull/8142/commits/35e3225f0478a7b952119dca2a800a76b9ae5f3e#diff-dbf04232abcedbe2e67996b043041b48) and it is upgraded to check (as suggested in comments) for used (custom) themes in products/categories/cms pages.

### Fixed Issues (if relevant)
1. magento/magento2#10363
2. magento/magento2#8145
3. magento/magento2#8142

### Manual testing scenarios
1. Add custom theme for product/category/cms page
2. Disable some theme
3. Upload product image
4. Check if cache image is generated for disabled theme

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
ADDED getThemesInUse (using [@vrann fix](https://github.com/magento/magento2/pull/8142/commits/35e3225f0478a7b952119dca2a800a76b9ae5f3e#diff-dbf04232abcedbe2e67996b043041b48) + added check for used themes in products, categories and cms pages)
REF magento2/issues#10363

UnitTests are failing due to mocking problems:
```
1) Magento\Catalog\Test\Unit\Model\Product\Image\CacheTest::testGenerate
Error: Call to a member function getId() on null
``

Commit:
Removed ObjectManager from Cache Class
Refactored Unit Test

ishakhsuvarov commented on app/code/Magento/Catalog/Model/Product/Image/Cache.php:
```diff
@@ -24,11 +27,27 @@ class Cache
     /**
+     * @var ThemeCustomizationConfig
+     */
+    protected $themeCustomizationConfig;
```
Adding new `protected` properties is expanding the contract of the class, which is not really desired in this case. Please change to `private`

denisristic replied:
changed to private

ishakhsuvarov commented on app/code/Magento/Catalog/Model/Product/Image/Cache.php:
```diff
@@ -127,4 +154,93 @@ protected function processImageData(Product $product, array $imageData, $file)
+     *
+     * @return array
+     */
+    public function getThemesInUse()
```
Does this method need to be public?
Please consider extracting it into appropriate service, as `getThemesInUse` does not seem to have too much in common with Product Image Cache Model.

denisristic replied:
Where would you suggest I should put it? Inside themes?

ishakhsuvarov commented on app/code/Magento/Catalog/Model/Product/Image/Cache.php:
```diff
@@ -127,4 +154,93 @@ protected function processImageData(Product $product, array $imageData, $file)
+            }
+        }
+
+        $connection = $this->attribute->getResource()->getConnection();
```
Getting dependencies in such way (one dependency via another one) violates the principle of least knowledge. 
Please consider refactoring this.

ishakhsuvarov commented on app/code/Magento/Catalog/Model/Product/Image/Cache.php:
```diff
@@ -127,4 +154,93 @@ protected function processImageData(Product $product, array $imageData, $file)
+
+        $productCustomDesignAttributeId = $this->attribute->loadByCode(4, 'custom_design')->getId();
+
+        $productSql = $connection
```
Direct access to the data layer is out of responsibilities scope for Magento Models. I suggest to move it into appropriate resource model. This would also help to resolve the commend above (on the least knowledge principle).

ishakhsuvarov commented on app/code/Magento/Catalog/Model/Product/Image/Cache.php:
```diff
@@ -127,4 +154,93 @@ protected function processImageData(Product $product, array $imageData, $file)
+
+        $connection = $this->attribute->getResource()->getConnection();
+
+        $productCustomDesignAttributeId = $this->attribute->loadByCode(4, 'custom_design')->getId();
```
Is there a way to avoid hardcoding the `4`?

denisristic replied:
fixed

ishakhsuvarov commented on app/code/Magento/Catalog/Model/Product/Image/Cache.php:
```diff
@@ -127,4 +154,93 @@ protected function processImageData(Product $product, array $imageData, $file)
+     *
+     * @return array
+     */
+    public function getThemesInUse()
```
Please add PHP 7.0 specific type hinting  and return types to any newly introduced methods.

denisristic replied:
added

Commit:
UPDATED as @ishakhsuvarov suggested
- Adding new protected properties is expanding the contract of the class, which is not really desired in this case. Please change to private
- Please add PHP 7.0 specific type hinting and return types to any newly introduced methods.
- Is there a way to avoid hardcoding the 4?

sidolov commented on app/code/Magento/Catalog/Model/Product/Image/Cache.php:
```diff
@@ -127,4 +154,97 @@ protected function processImageData(Product $product, array $imageData, $file)
+
+        $productEntityTypeId = \Magento\Catalog\Setup\CategorySetup::CATALOG_PRODUCT_ENTITY_TYPE_ID;
+
+        $productCustomDesignAttributeId = $this->attribute->loadByCode($productEntityTypeId, 'custom_design')->getId();
```
Please use \Magento\Eav\Model\Config::getAttribute instead of load entity by model

sidolov commented on app/code/Magento/Catalog/Model/Product/Image/Cache.php:
```diff
@@ -127,4 +154,97 @@ protected function processImageData(Product $product, array $imageData, $file)
+
+        $connection = $this->attribute->getResource()->getConnection();
+
+        $productEntityTypeId = \Magento\Catalog\Setup\CategorySetup::CATALOG_PRODUCT_ENTITY_TYPE_ID;
```
Please use constant directly instead of assign to variable

sidolov commented on app/code/Magento/Catalog/Model/Product/Image/Cache.php:
```diff
@@ -41,11 +60,17 @@ class Cache
         ThemeCollection $themeCollection,
-        ImageHelper $imageHelper
+        ImageHelper $imageHelper,
+        ThemeCustomizationConfig $themeCustomizationConfig,
```
Dependencies added with backward compatibility standards violation, right way described in http://devdocs.magento.com/guides/v2.2/contributor-guide/backward-compatible-development/ section: Adding a constructor parameter

sidolov commented on app/code/Magento/Catalog/Model/Product/Image/Cache.php:
```diff
@@ -127,4 +154,97 @@ protected function processImageData(Product $product, array $imageData, $file)
+        $storesByThemes   = $this->themeCustomizationConfig->getStoresByThemes();
+
+        foreach ($registeredThemes as $registeredTheme) {
+            if (array_key_exists($registeredTheme->getThemeId(), $storesByThemes)) {
```
Use $registeredTheme->getCode() instead of $registeredTheme->getThemeId() otherwise this expression will return false permanently and $themesInUse array will be always empty. Without this fix cache directory is empty.

sidolov commented on app/code/Magento/Catalog/Model/Product/Image/Cache.php:
```diff
@@ -127,4 +154,97 @@ protected function processImageData(Product $product, array $imageData, $file)
+     *
+     * @return array
+     */
+    private function getThemesInUse(): array
```
Consider refactoring this logic with introducing composite object (with configuration via DI) for themes in use and move logic to corresponded modules (pages to CMS module; product, category -> Catalog module etc)

sidolov commented on app/code/Magento/Catalog/Model/Product/Image/Cache.php:
```diff
@@ -23,11 +26,27 @@ class Cache
+    /**
+     * @var Config
+     */
+    protected $eavConfig;
```
Please use private modifier for object dependencies

sidolov commented on app/code/Magento/Catalog/Model/Product/Image/Cache.php:
```diff
@@ -23,11 +26,27 @@ class Cache
+    /**
+     * @var Attribute
+     */
+    protected $attribute;
```
Please use private modifier for object dependencies

sidolov commented on app/code/Magento/Catalog/Model/Product/Image/Cache.php:
```diff
@@ -127,4 +154,97 @@ protected function processImageData(Product $product, array $imageData, $file)
+            }
+        }
+
+        $categoryEntityTypeId = \Magento\Catalog\Setup\CategorySetup::CATEGORY_ENTITY_TYPE_ID;
```
Please use constant directly instead of assign to variable

sidolov commented on app/code/Magento/Catalog/Model/Product/Image/Cache.php:
```diff
@@ -127,4 +154,97 @@ protected function processImageData(Product $product, array $imageData, $file)
+
+        $categoryEntityTypeId = \Magento\Catalog\Setup\CategorySetup::CATEGORY_ENTITY_TYPE_ID;
+
+        $categoryCustomDesignAttributeId = $this->attribute->loadByCode($categoryEntityTypeId, 'custom_design')->getId();
```
Please use \Magento\Eav\Model\Config::getAttribute instead of load entity by model

sidolov commented:
@denisristic,  I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for collaboration
